src/axl.c

```diff
@@ -309,6 +310,124 @@ int AXL_Finalize (void)
+    return NULL;
+  }
+
+  if (! configured) {
+    if (config != NULL) {
+      /* dummy variables for options not actually supported (anymore) */
+      double axl_flush_async_bw = 0.0;
+      double axl_flush_async_percent = 0.0;
+      int axl_crc_on_flush = 1;
+
+      /* read out all options we know about */
+      /* TODO: this could be turned into a list of structs */
+      unsigned long ul;
+      if (kvtree_util_get_bytecount(config,
+        AXL_KEY_CONFIG_FLUSH_ASYNC_BW, &ul) == KVTREE_SUCCESS)
+      {
+        axl_flush_async_bw = (double) ul;
+        if (axl_flush_async_bw != ul) {
+          char* value;
+          kvtree_util_get_str(config, AXL_KEY_CONFIG_FLUSH_ASYNC_BW, &value);
+          AXL_ERR("Value '%s' passed for %s exceeds int range",
+            value, AXL_KEY_CONFIG_FLUSH_ASYNC_BW
+          );
+          retval = NULL;
+        }
+      }
+
+      kvtree_util_get_double(config,
+        AXL_KEY_CONFIG_FLUSH_ASYNC_PERCENT, &axl_flush_async_percent);
+
+      if (kvtree_util_get_bytecount(config,
+        AXL_KEY_CONFIG_FILE_BUF_SIZE, &ul) == KVTREE_SUCCESS)
+      {
+        axl_file_buf_size = (int) ul;
+        if (axl_file_buf_size != ul) {
+          char* value;
+          kvtree_util_get_str(config, AXL_KEY_CONFIG_FILE_BUF_SIZE, &value);
+          AXL_ERR("Value '%s' passed for %s exceeds int range",
+            value, AXL_KEY_CONFIG_FILE_BUF_SIZE
+          );
+          retval = NULL;
+        }
+      }
+
+      kvtree_util_get_int(config,
+        AXL_KEY_CONFIG_CRC_ON_FLUSH, &axl_crc_on_flush);
+
+      kvtree_util_get_int(config, AXL_KEY_CONFIG_DEBUG, &axl_debug);
+
+      kvtree_util_get_int(config, AXL_KEY_CONFIG_MKDIR, &axl_make_directories);
+
+      kvtree_util_get_int(config,
+        AXL_KEY_CONFIG_COPY_METADATA, &axl_copy_metadata);
+
+      /* report all unknown options (typos?) */
+      const kvtree_elem* elem;
+      for (elem = kvtree_elem_first(config);
+           elem != NULL;
+           elem = kvtree_elem_next(elem))
+      {
+        /* must be only one level deep, ie plain kev = value */
+        const kvtree* elem_hash = kvtree_elem_hash(elem);
+        assert(kvtree_size(elem_hash) == 1);
+
+        const kvtree* kvtree_first_elem_hash =
+          kvtree_elem_hash(kvtree_elem_first(elem_hash));
+        assert(kvtree_size(kvtree_first_elem_hash) == 0);
+
+        /* check against known options */
+        const char** opt;
+        int found = 0;
+        for (opt = known_options; *opt != NULL; opt++) {
+          if (strcmp(*opt, kvtree_elem_key(elem)) == 0) {
+            found = 1;
+            break;
+          }
+        }
+        if (! found) {
+          AXL_ERR("Unknown configuration parameter '%s' with value '%s'",
+            kvtree_elem_key(elem),
+            kvtree_elem_key(kvtree_elem_first(kvtree_elem_hash(elem)))
+          );
+          retval = NULL;
+        }
+      }
+    }
+
+    /* only accept configuration options once */
+    configured = 1;
```

Comment:
Regarding the "only accept configuration options once", you had earlier mentioned:

> Right now the check is in place to reduce the chance of setting parameters, which AXL currently can assume to be constant during a run, to change in the middle of a run (at least to the amount that this is possible given that the config calls happen after AXL_Init()

I don't think we can make that assumption, since we can't predict what the user will want to do, nor what config options we may want to add in the future.  It also prevents the user from setting different config options at different spots in their code.   For example, a user couldn't set AXL_KEY_CONFIG_DEBUG in their "init" code and then later set  AXL_KEY_CONFIG_MKDIR in their "parsing my command line options" code.  My vote would be to let the user call AXL_Config() as many times as they want.

Reply:
I changed the code as suggested.

Reply:
This (changing option at runtime in particular after having created a transfer) is affected by #85 since AXL also accesses the global values of the configuration options without protecting them between multiple threads. Right now the best solution seems to me to make copies of the global variables when a new transfer is created and initialize the transfer-local copies with the global values. This way there is consistent behaviour when changing configuration options at runtime namely a newly created AXL transfer will use the global options active at the time it was created (rather than whatever value the global options had at the time the thread driving the transfer reads them out).

Reply:
> Right now the best solution seems to me to make copies of the global variables when a new transfer is created and initialize the transfer-local copies with the global values. 

Then they don't behave as globals anymore.   A better solution is to protect the globals with a reader-writer lock.

Reply:
There seems to be a misunderstanding. I am proposing to initialize the per-transfer options with the values of the global options at the time the transfer is created. AXL in master will always use the current value of the global variables. This pull request makes these global variables that hold the configuration values non-const in between `AXL_Init` and the end of the code which means the observed behaviour is that the behaviour of a transfer is a race condition between when the transfer thread looks at the global variables and when the user code calls `AXL_Config`. 
A possible way around, that I had not considered before, is to forbid calls to `AXL_Config` after the first transfer has been started thought that does not seem to be what was requested.

The global variables will need to be protected by a lock if they are used in the threads, which is, at the very least potentially true for the `axl_debug` option. This would be in addition to making a copy of their values when creating the transfer. Simiarly the function that modify per-transfer options need to get a lock before they modify the transfer options and the threads using them need to use the same lock.

I am admitting to confusion why a global option not being constant would make it not global anymore.

Reply:
>  I am proposing to initialize the per-transfer options with the values of the global options at the time the transfer is created. 

Sorry, I'm confused.  Are you saying the defaults for the per-transfer config values would be configurable as globals?  For example, having a configurable global `file_buf_size` which the per-transfer `file_buf_size` values get initially set to?

Reply:
Well mostly. I am not really saying "would be". I am saying "this is what AXL does in its master branch". Namely there is a global variable `axl_file_buf_size` which is initialized to `(size_t) 1048576` in `__AXL_Init` (and written to only there, it is not changed anywhere else). 

This global variable is then used later on eg as an argument to `axl_file_copy`: `axl_file_copy(source, destination, axl_file_buf_size, resume)` by all transfers.

The configuration interface will change this global variable if the "global" ID is passed to it.

Thus when allowing per transfer configurable `file_buf_size` it seem natural that if a user does not explicitly configure `file_buf_size` then the value of `axl_file_buf_size` is used since this is the behaviour in `master` (where no configuration is possible). 

Does this help clarify what I tried to say?

Reply:
Looks good to me.  Until we modify the code to change config settings on a per transfer basis, this lets the AXL user to change hard coded values.  Per transfer settings is another level up and can be added in a later PR.

Reply:
@adammoody that's fine, we can tack on per-transfer config options later.